lib/vm/src/memory.rs

```diff
@@ -7,53 +8,65 @@ use cosmwasm::memory::Slice;
 // write_mem allocates memory in the instance and copies the given data in
 // returns the memory offset, to be passed as an argument
 // panics on any error (TODO, use result?)
-pub fn allocate(instance: &mut Instance, data: &[u8]) -> i32 {
+pub fn allocate(instance: &mut Instance, data: &[u8]) -> u32 {
     // allocate
-    let alloc: Func<(i32), (i32)> = instance.func("allocate").unwrap();
-    let ptr = alloc.call(data.len() as i32).unwrap();
+    let alloc: Func<(u32), (u32)> = instance.func("allocate").unwrap();
+    let ptr = alloc.call(data.len() as u32).unwrap();
     write_memory(instance.context(), ptr, data);
     ptr
 }
 
-pub fn read_memory(ctx: &Ctx, ptr: i32) -> Vec<u8> {
+pub fn read_memory(ctx: &Ctx, ptr: u32) -> Vec<u8> {
     let slice = to_slice(ctx, ptr);
-    let (start, end) = (slice.offset, slice.offset + slice.len);
-    let memory = &ctx.memory(0).view::<u8>()[start..end];
+    let memory = ctx.memory(0);
+    let len = slice.len as usize;
+    let mut result = vec![0u8; len];
 
     // TODO: there must be a faster way to copy memory
-    let mut result = vec![0u8; slice.len];
-    for i in 0..slice.len {
-        result[i] = memory[i].get();
+    let buffer = WasmPtr::<u8, Array>::new(slice.offset)
+        .deref(memory, 0, slice.len)
+        .unwrap();
+    for i in 0..len {
+        result[i] = buffer[i].get();
     }
     result
 }
 
 // write_memory returns how many bytes written on success
 // negative result is how many bytes requested if too small
-pub fn write_memory(ctx: &Ctx, ptr: i32, data: &[u8]) -> i32 {
+pub fn write_memory(ctx: &Ctx, ptr: u32, data: &[u8]) -> i32 {
     let slice = to_slice(ctx, ptr);
-    if data.len() > slice.len {
+    if data.len() > (slice.len as usize) {
         return -(data.len() as i32);
     }
     if data.len() == 0 {
         return 0;
     }
 
-    let (start, end) = (slice.offset, slice.offset + slice.len);
-    let memory = &ctx.memory(0).view::<u8>()[start..end];
+    let memory = ctx.memory(0);
     // TODO: there must be a faster way to copy memory
+    let buffer = unsafe {
+        WasmPtr::<u8, Array>::new(slice.offset)
+            .deref_mut(memory, 0, slice.len)
+            .unwrap()
+    };
     for i in 0..data.len() {
-        memory[i].set(data[i])
+        buffer[i].set(data[i])
     }
     data.len() as i32
 }
 
 // to_slice reads in a ptr to slice in wasm memory and constructs the object we can use to access it
-fn to_slice(ctx: &Ctx, ptr: i32) -> Slice {
-    let buf_ptr = (ptr / 4) as usize; // convert from u8 to i32 offset
-    let memory = &ctx.memory(0).view::<i32>();
+fn to_slice(ctx: &Ctx, ptr: u32) -> Slice {
```

Comment:
Because `WasmPtr` is bit-for-bit identical to u32, you can use it in arguments here

So you can write this as:

```Rust
#[repr(C)]
#[derive(Default)]
struct MySlice {
   ptr: u32, // `WasmPtr<T, Array>` is also valid here too!
   len: u32,
}

// probably necessary, this is how we check that it's safe to convert values to and from LE bytes
unsafe impl wasmer_runtime_core::types::ValueType for MySlice { }

fn to_slice(ctx: &Ctx, ptr: WasmPtr<MySlice>) -> MySlice {
   let memory = &ctx.memory(0);
   ptr.deref(memory).map(|x| x.get()).unwrap_or_default()
}
```

Reply:
Very nice tip, I will do.
I had started to use `WasmPtr<u32>` as an arg here, but then didn't know how to increment it. I like this way, actually make full use of `repr(C)`.

Yesterday, I used `usize`, which would be different in host and wasm, but I modified to `u32` for `WasmPtr`, which actually does let this work. I will update, much cleaner.

Reply:
Re: `// `WasmPtr<T, Array>` is also valid here too!`

Ah, so they also show up as u32 for repr(C)... cool recursion.
But...I want to use this code in both sides of the vm boundary - in the wasm code and in the calling code. I am do my best to avoid any imports of wasmer inside the wasm code, keeping it as trim as possible. Otherwise, I would try this.

Reply:
Yeah, that's fine!  If you have a type with a lot of these, you can always define a zero-cost conversion between them.

I think the trick here is that WasmPtr is marked as `#[repr(transparent)]` and it uses `PhantomData` to store the type information and whether it's an Array or not.  `PhantomData` is compile-time only and `repr(transparent)` means that the struct acts as an invisible wrapper around its single inner type.  Put another way, it makes your struct ABI compatible with its inner type; which is why it works here.